internal/caching/expiring_cache.go

```diff
@@ -0,0 +1,149 @@
+package caching
+
+import (
+	"errors"
+	"reflect"
+	"sync"
+	"time"
+)
+
+var (
+	ErrNoCacheData = errors.New("no cache data available")
+	ErrStaleData   = errors.New("stale cache data")
+)
+
+type cachedEntry[T any] struct {
+	data      T
+	timestamp time.Time
+	stale     bool
+}
+
+// Cache provides a generic caching mechanism with validity checking
+type Cache[T any] struct {
+	entry     *cachedEntry[T]
+	ttl       time.Duration
+	fetchFunc func() (T, error)
+	mutex     sync.RWMutex
+}
+
+// NewCacheWithTTL creates a new cache with a custom validity period and fetch function
+func NewCacheWithTTL[T any](validity time.Duration, fetchFunc func() (T, error)) *Cache[T] {
+	return &Cache[T]{
+		ttl:       validity,
+		fetchFunc: fetchFunc,
+	}
+}
+
+// Get returns the data either from cache or fetches from origin depending on validity period
+func (c *Cache[T]) Get() (T, error) {
```

Comment:
Do we have performance issues when using just write lock for the whole `Get` or there is other reason to have double locking here?